File: algoritmos de treino/outros/tabela_hash.rb
# in order to be accessible from within the check_voter function
# voted has to be an instance variable (or global / class)
@voted = {}

def check_voter(name)
  if @voted[name]
    puts "kick them out!"
  else
    @voted[name] = true
    puts "let them vote!"
  end
end

check_voter("tom")
check_voter("mike")
check_voter("mike")

# {} is the same as Hash.new
#book = {}
# an apple costs 67 cents
#book["apple"] = 0.67
# milk costs $1.49
#book["milk"] = 1.49
#book["avocado"] = 1.49
#puts book

File: algoritmos de treino/outros/tabelaHash.cpp
#include <iostream>
#include <unordered_map>
#include <string>

using std::cout;
using std::endl;

std::unordered_map<std::string, bool> voted;

void check_voter(const std::string& name) {
    auto search = voted.find(name);
    if (search == voted.end() || search->second == false) {
        voted.insert({name, true});
        cout << "Let them vote!" << endl;;
    } else {
        cout << "Kick them out!" << endl;
    }
}

int main() {
    check_voter("tom");
    check_voter("mike");
    check_voter("mike");
}

/*#include <iostream>
#include <unordered_map>
#include <string>
#include <utility>

using std::cout;
using std::endl;

int main() {
    std::unordered_map<std::string, float> book = {
        {"apple", 0.67},
        {"milk", 1.49},
        {"avocado", 1.49}
    };

    // print book
    for (std::pair<std::string, float> pair : book) {
        cout << pair.first << ": " << pair.second << "$" << endl;
    }
}*/